docs/src/tutorials/nonlinear/user_defined_hessians.jl

```diff
@@ -0,0 +1,210 @@
+# Permission is hereby granted, free of charge, to any person obtaining a copy   #src
+# of this software and associated documentation files (the "Software"), to deal  #src
+# in the Software without restriction, including without limitation the rights   #src
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell      #src
+# copies of the Software, and to permit persons to whom the Software is          #src
+# furnished to do so, subject to the following conditions:                       #src
+#                                                                                #src
+# The above copyright notice and this permission notice shall be included in all #src
+# copies or substantial portions of the Software.                                #src
+#                                                                                #src
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR     #src
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,       #src
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE    #src
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER         #src
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,  #src
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE  #src
+# SOFTWARE.                                                                      #src
+
+# # User-defined Hessians
+
+# In this tutorial, we explain how to write a user-defined function (see [User-defined Functions](@ref))
+# with an explicit Hessian matrix.
+
+# This tutorial uses the following packages:
+
+using JuMP
+import Ipopt
+
+# ## Motivation
+
+# Our goal for this tutorial is to solve the bilevel optimization problem:
+
+# ```math
+# \begin{array}{r l}
+# \min\limits_{x} & x_1^2 + x_2^2 + z \\
+# s.t.            & \begin{array}{r l}
+#                       z \ge \max\limits_{y} & x_1^2 y_1 + x_2^2 y_2  - x_1 y_1^4 - 2 x_2 y_2^4 \\
+#                       s.t.                  & (y_1 - 10)^2 + (y_2 - 10)^2 \le 25
+#                   \end{array} \\
+#                 & x \ge 0.
+# \end{array}
+# ```
+
+# This bilevel optimization problem is composed of two nested optimization
+# problems. An _upper_ level, involving variables ``x``, and a _lower_ level,
+# involving variables ``y``. From the perspective of the lower-level, the
+# values of ``x`` are fixed parameters, and so the model optimizes ``y`` given
+# those fixed parameters. Simultaneously, the upper level is optimizing ``x``
+# given the response of ``y``.
+
+# ## Decomposition
+
+# There are a few ways to solve this problem, but we are going to use a
+# nonlinear decomposition method. The first step is to write a function to
+# compute:
+
+# ```math
+# \begin{array}{r l}
+#   V(x_1, x_z) = \max\limits_{y} & x_1^2 y_1 + x_2^2 y_2  - x_1 y_1^4 - 2 x_2 y_2^4 \\
```

Comment:
x_z seems like a typo, should it be x_2?

Reply:
I must be going blind. That's a nasty one to spot.